File: packages/sdk-go/pricing_test.go
package rateguard

import "testing"

func TestNormalizeModelID(t *testing.T) {
	cases := map[string]string{
		"gpt-4o-2024-08-06":        "gpt-4o",                // OpenAI ISO snapshot
		"gpt-4.1-2025-04-14":       "gpt-4.1",               // OpenAI, dotted version kept
		"o3-2025-04-16":            "o3",                    // OpenAI reasoning snapshot
		"claude-sonnet-4-20250514": "claude-sonnet-4",       // Anthropic compact date
		"claude-opus-4-5-20251101": "claude-opus-4-5",       // Anthropic date, minor kept
		"gemini-2.5-flash-09-2025": "gemini-2.5-flash",      // Gemini MM-YYYY
		"gemini-2.5-flash-preview": "gemini-2.5-flash",      // Gemini preview alias
		"gemini-2.5-flash-latest":  "gemini-2.5-flash",      // Gemini latest alias
		"GPT-4O":                   "gpt-4o",                // case-folded
		"gpt-4o-mini":              "gpt-4o-mini",           // meaningful word NOT stripped
		"o4-mini":                  "o4-mini",               // meaningful word NOT stripped
		"claude-sonnet-4-5":        "claude-sonnet-4-5",     // bare minor version NOT stripped
		"gemini-2.5-flash-lite":    "gemini-2.5-flash-lite", // lite NOT stripped
		"my-custom-finetune":       "my-custom-finetune",    // unknown left intact
	}
	for in, want := range cases {
		if got := normalizeModelID(in); got != want {
			t.Errorf("normalizeModelID(%q) = %q, want %q", in, got, want)
		}
	}
}

func TestBuiltinPricingMatchesDatedSnapshot(t *testing.T) {
	// A provider-reported dated ID must resolve to the base table entry —
	// otherwise every real streaming response prices at $0.
	bare := EstimateCost("gpt-4o", 1000, 1000)
	dated := EstimateCost("gpt-4o-2024-08-06", 1000, 1000)
	if bare == 0 {
		t.Fatal("gpt-4o should have a built-in price")
	}
	if dated != bare {
		t.Errorf("dated snapshot priced %v, base priced %v — normalization not applied", dated, bare)
	}
}

func TestStaticPricingProviderOverridesAndFallsThrough(t *testing.T) {
	p := StaticPricing{
		"my-model": {PromptUSDPer1K: 0.001, CompletionUSDPer1K: 0.002},
		"gpt-4o":   {PromptUSDPer1K: 1.0, CompletionUSDPer1K: 2.0}, // override the built-in
	}

	// Custom model the built-in table has never heard of.
	if got := estimateCostWith(p, "my-model", 1000, 1000); got != 0.001+0.002 {
		t.Errorf("custom model cost = %v, want %v", got, 0.003)
	}
	// User override wins over the built-in table.
	if got := estimateCostWith(p, "gpt-4o", 1000, 1000); got != 1.0+2.0 {
		t.Errorf("override cost = %v, want 3.0", got)
	}
	// A dated snapshot of the overridden model still resolves via normalization.
	if got := estimateCostWith(p, "gpt-4o-2024-08-06", 1000, 1000); got != 3.0 {
		t.Errorf("dated override cost = %v, want 3.0", got)
	}
	// Provider miss falls through to the built-in table.
	if got := estimateCostWith(p, "claude-sonnet-4", 1000, 1000); got == 0 {
		t.Error("provider miss should fall through to the built-in table, got 0")
	}
	// Genuinely unknown everywhere → zero, never fabricated.
	if got := estimateCostWith(p, "totally-unknown-model", 1000, 1000); got != 0 {
		t.Errorf("unknown model cost = %v, want 0", got)
	}
}

File: packages/sdk-node/test/pricing.test.ts
import { describe, it, expect } from 'vitest';
import { estimateCost, estimateCostWith, normalizeModelId, StaticPricing } from '../src/core/genai.js';

describe('normalizeModelId', () => {
  const cases: Record<string, string> = {
    'gpt-4o-2024-08-06': 'gpt-4o', // OpenAI ISO snapshot
    'gpt-4.1-2025-04-14': 'gpt-4.1', // dotted version kept
    'o3-2025-04-16': 'o3',
    'claude-sonnet-4-20250514': 'claude-sonnet-4', // Anthropic compact date
    'claude-opus-4-5-20251101': 'claude-opus-4-5', // date stripped, minor kept
    'gemini-2.5-flash-09-2025': 'gemini-2.5-flash', // MM-YYYY
    'gemini-2.5-flash-preview': 'gemini-2.5-flash',
    'gemini-2.5-flash-latest': 'gemini-2.5-flash',
    'GPT-4O': 'gpt-4o', // case-folded
    'gpt-4o-mini': 'gpt-4o-mini', // meaningful word NOT stripped
    'o4-mini': 'o4-mini',
    'claude-sonnet-4-5': 'claude-sonnet-4-5', // bare minor version NOT stripped
    'gemini-2.5-flash-lite': 'gemini-2.5-flash-lite', // lite NOT stripped
    'my-custom-finetune': 'my-custom-finetune',
  };
  for (const [input, want] of Object.entries(cases)) {
    it(`normalizes ${input} -> ${want}`, () => {
      expect(normalizeModelId(input)).toBe(want);
    });
  }
});

describe('pricing', () => {
  it('matches a dated snapshot to its base table entry (not $0)', () => {
    const bare = estimateCost('gpt-4o', 1000, 1000);
    expect(bare).toBeGreaterThan(0);
    expect(estimateCost('gpt-4o-2024-08-06', 1000, 1000)).toBe(bare);
  });

  it('a StaticPricing provider overrides, normalizes, and falls through', () => {
    const p = new StaticPricing({
      'my-model': { promptUSDPer1K: 0.001, completionUSDPer1K: 0.002 },
      'gpt-4o': { promptUSDPer1K: 1.0, completionUSDPer1K: 2.0 }, // override built-in
    });
    // custom model the table has never heard of
    expect(estimateCostWith(p, 'my-model', 1000, 1000)).toBeCloseTo(0.003);
    // override wins over the built-in table
    expect(estimateCostWith(p, 'gpt-4o', 1000, 1000)).toBeCloseTo(3.0);
    // dated snapshot of the overridden model resolves via normalization
    expect(estimateCostWith(p, 'gpt-4o-2024-08-06', 1000, 1000)).toBeCloseTo(3.0);
    // provider miss falls through to the built-in table
    expect(estimateCostWith(p, 'claude-sonnet-4', 1000, 1000)).toBeGreaterThan(0);
    // unknown everywhere -> zero, never fabricated
    expect(estimateCostWith(p, 'totally-unknown-model', 1000, 1000)).toBe(0);
  });
});
